File: algorithms/1_easy/L/02373_largest_local_values_in_a_matrix/solution.cpp
class Solution {
public:
    static vector<vector<int>> largestLocal(const vector<vector<int>>& grid) {
        const size_t n = grid.size();
        assert(n >= 3 && n == grid[0].size());
        auto res =  vector<vector<int>>();
        res.reserve(n - 2);

        for (size_t i = 0; i < n - 2; i += 1) {
            auto tmp = vector<int>();
            tmp.reserve(n - 2);

            for (size_t j = 0; j < n - 2; j += 1) {
                int max = numeric_limits<int>::min();
                for (size_t ofs_i = 0; ofs_i < 3; ofs_i += 1) {
                    for (size_t ofs_j = 0; ofs_j < 3; ofs_j += 1) {
                        max = std::max(max, grid[i + ofs_i][j + ofs_j]);
                    }
                }
                tmp.push_back(max);
            }

            res.push_back(tmp);
        }

        return res;
    }
};

File: algorithms/1_easy/L/02373_largest_local_values_in_a_matrix/solution.rs
impl Solution {
    pub fn largest_local(grid: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let n = grid.len();
        assert!(n >= 3 && n == grid[0].len());
        let mut res: Vec<Vec<i32>> = Vec::with_capacity(n - 2);

        for i in 0..(n - 2) {
            let mut tmp: Vec<i32> = Vec::with_capacity(n - 2);

            for j in 0..(n - 2) {
                let mut max = i32::MIN;
                for ofs_i in 0..3 {
                    for ofs_j in 0..3 {
                        max = max.max(grid[i + ofs_i][j + ofs_j]);
                    }
                }
                tmp.push(max);
            }

            res.push(tmp);
        }

        res
    }
}
